src/protocols/secure_channel/CASEServer.cpp

```diff
@@ -146,7 +157,16 @@ void CASEServer::PrepareForSessionEstablishment(const ScopedNodeId & previouslyE
 void CASEServer::OnSessionEstablishmentError(CHIP_ERROR err)
 {
     ChipLogError(Inet, "CASE Session establishment failed: %s", ErrorStr(err));
-    PrepareForSessionEstablishment();
+
+    //
+    // We're not allowed to call methods that will eventually result in calling SessionManager::AllocateSecureSession
+    // from a SessionDelegate::OnSessionReleased callback. Schedule the preparation as an async work item.
+    //
+    mSessionManager->SystemLayer()->ScheduleWork([](auto *systemLayer, auto *appState) {
+        CASEServer *_this = static_cast<CASEServer *>(appState);
+        _this->PrepareForSessionEstablishment();
```

Comment:
Can you say more about why this restriction exists?  What is the implication of posting this to the work queue for corner cases like this occurring immediately before stack shutdown?

Reply:
When running the session eviction algorithm in `SecureSessionTable::EvictAndAllocate`, we have to call `MarkForEviction` on a session that has been selected. This in turn, will detach all attached `SessionHolder`s, and call their `OnSessionReleased` method. If that in turn calls `AllocateSession`, that's going to end up back again in `SecureSessionTable::EvictAndAllocate`, which isn't re-entrant.

A big part of the problem here is that don't know apriori for a given candidate session, if calling `MarkForEviction` is going to actually result in a session with no remaining references to it. All that does is kick off the holders. If we correctly pivot SessionHolder to be a true weak-reference, then we would not need this speculative eviction logic, since we can always tell if there are true strong-refs attached, and skip to a session that we know for sure we will be free. If we did that, we _can_ make this method re-entrant...

Reply:
> What is the implication of posting this to the work queue for corner cases like this occurring immediately before stack shutdown?

I don't see any particular problem there, since if the stack is about to shut-down, either it will process all remaining work queue items (and process this posted event), or purge them entirely (though I doubt that's the behavior today).

Do you see a specific problem here?